File: src/#117/solution.go
package solution

type Node struct {
	Val               int
	Left, Right, Next *Node
}

func connect(root *Node) *Node {
	levelMap := map[int]*Node{}
	helper(root, levelMap, 0)
	return root
}

func helper(node *Node, levelMap map[int]*Node, level int) {
	if node == nil {
		return
	}
	if _, ok := levelMap[level]; ok {
		levelMap[level].Next = node
		levelMap[level] = node
	} else {
		levelMap[level] = node
	}
	helper(node.Left, levelMap, level+1)
	helper(node.Right, levelMap, level+1)

}

File: src/#117/solution.js
function Node(val, left, right, next) {
  this.val = val === undefined ? null : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
  this.next = next === undefined ? null : next
}

var connect = function (root) {
  if (!root) return root
  let queue = [root]

  while (queue.length) {
    let num = queue.length

    for (let i = 0; i < num; i++) {
      let curr = queue.shift()

      if (i !== num - 1) {
        curr.next = queue[0]
      }

      if (curr.left) queue.push(curr.left)
      if (curr.right) queue.push(curr.right)
    }
  }
  return root
}
